File: cpp/Stack_Queue/linkedlist_stack.cpp
#include <stdio.h>
#include <vector>

using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int val) : val(val), next(nullptr) {}
}

/* Stack class based on linked list */
class LinkedListStack
{
private:
    ListNode *stackTop; // Use the head node as the top of the stack
    int stkSize;        // Length of the stack

public:
    LinkedListStack()
    {
        stackTop = nullptr;
        stkSize = 0;
    }

    ~LinkedListStack()
    {
        // Traverse the linked list, remove nodes, free memory
        freeMemoryLinkedList(stackTop);
    }

    /* Get the length of the stack */
    int size()
    {
        return stkSize;
    }

    /* Determine if the stack is empty */
    bool isEmpty()
    {
        return size() == 0;
    }

    /* Push */
    void push(int num)
    {
        ListNode *node = new ListNode(num);
        node->next = stackTop;
        stackTop = node;
        stkSize++;
    }

    /* Pop */
    int pop()
    {
        int num = top();
        ListNode *tmp = stackTop;
        stackTop = stackTop->next;
        // Free memory
        delete tmp;
        stkSize--;
        return num;
    }

    /* Access stack top element */
    int top()
    {
        if (isEmpty())
            throw out_of_range("Stack is empty");
        return stackTop->val;
    }

    /* Convert the List to Array and return */
    vector<int> toVector()
    {
        ListNode *node = stackTop;
        vector<int> res(size());
        for (int i = res.size() - 1; i >= 0; i--)
        {
            res[i] = node->val;
            node = node->next;
        }
        return res;
    }
};

int main(int argc, char const *argv[])
{
    LinkedListStack stack;
    stack.push(1);
    stack.push(3);
    stack.push(2);
    stack.push(5);
    stack.push(4);

    vector<int> res = stack.toVector();
    for (int i = 0; i < res.size(); i++)
    {
        printf("%d ", res[i]);
    }

    return 0;
}

File: java/stack/linkedlist_stack.java
/* Linked list node class */
class ListNode {
  int val; // Node value
  ListNode next; // Reference to next node

  ListNode(int x) { // Constructor
    val = x;
  }
}

/* Stack class based on linked list */
class LinkedListStack {
  private ListNode stackPeek; // Use the head node as the top of the stack
  private int stkSize = 0; // Length of the stack

  public LinkedListStack() {
    stackPeek = null;
  }

  /* Get the length of the stack */
  public int size() {
    return stkSize;
  }

  /* Determine if the stack is empty */
  public boolean isEmpty() {
    return size() == 0;
  }

  /* Push */
  public void push(int num) {
    ListNode node = new ListNode(num);
    node.next = stackPeek;
    stackPeek = node;
    stkSize++;
  }

  /* Pop */
  public int pop() {
    int num = peek();
    stackPeek = stackPeek.next;
    stkSize--;
    return num;
  }

  /* Access stack top element */
  public int peek() {
    if (isEmpty()) throw new IndexOutOfBoundsException();
    return stackPeek.val;
  }

  /* Convert the List to Array and return */
  public int[] toArray() {
    ListNode node = stackPeek;
    int[] res = new int[size()];
    for (int i = res.length - 1; i >= 0; i--) {
      res[i] = node.val;
      node = node.next;
    }
    return res;
  }
}
